### PR TITLE
Resolve and replace commands for next command

### LINKED ISSUE #89758: pipeTransport.pipeArgs command variable replacement not working
Attempting to figure out why input and command replacements do not work for `launch.json` `pipeTransport.pipeArgs` (used in omnisharp-csharp extension).  My use case requires `omnisharp-vscode` extension.  However, this involves the vscode variable replacement logic, so I am posting it here.

- VSCode Version: `1.41.1` (also tried insider `1.42.0-insider`)
- OS Version: `Darwin x64 18.7.0`

Steps to Reproduce:

1. Create a `launch.json` for remote debugging a csharp docker container. (e.g. `vsdbg`)

**My Use Case Example** (requires `vscode-docker` extension)
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Docker Attach",
            "type": "coreclr",
            "request": "attach",
            "pipeTransport": {
                "pipeProgram": "docker",
                "pipeArgs": [
                    "exec", "-i", "${command:vscode-docker.containers.select}"
                ],
                "debuggerPath": "/root/vsdbg/vsdbg",
                "pipeCwd": "${workspaceRoot}",
                "quoteArgs": false
            },
            "processId": "${command:pickRemoteProcess}",
            "sourceFileMap": {
                "/app": "${workspaceRoot}"
            },
            "justMyCode": true
        }
    ]
}
```
vscode-docker.containers.select reference: https://github.com/microsoft/vscode-docker/pull/1240

**Another Example**
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Docker Attach",
            "type": "coreclr",
            "request": "attach",
            "pipeTransport": {
                "pipeProgram": "docker",
                "pipeArgs": [
                    "exec", "-i", "${input:sampleInput}"
                ],
                "debuggerPath": "/root/vsdbg/vsdbg",
                "pipeCwd": "${workspaceRoot}",
                "quoteArgs": false
            },
            "processId": "${command:pickRemoteProcess}",
            "sourceFileMap": {
                "/app": "${workspaceRoot}"
            },
            "justMyCode": true
        }
    ],
    "inputs": [
        {
            "id": "sampleInput",
            "type": "promptString",
            "description": "issue example"
        }
    ]
}
```

2. It will display the vscode prompts and dialogs, but not actually replace the entered/command value into the string for either example.

**Output**
```
Executing: docker exec -i ${input:sampleInput} sh -s < /Users/user/.vscode/extensions/ms-vscode.csharp-1.21.9/scripts/remoteProcessPickerScript
stderr: "docker exec" requires at least 2 arguments.
...
```

Does this issue occur when all extensions are disabled?: **Not Sure.**

### PR DESCRIPTION
**Pull request by davidruhmann**
This PR fixes #89758

Which will add support for the use case of using command and environment values in the configuration as they are resolved for subsequent command executions.  This enables the following use case.
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Docker Attach",
            "type": "coreclr",
            "request": "attach",
            "pipeTransport": {
                "pipeProgram": "docker",
                "pipeArgs": [
                    "exec", "-i", "${command:vscode-docker.containers.select}"
                ],
                "debuggerPath": "/root/vsdbg/vsdbg",
                "pipeCwd": "${workspaceRoot}",
                "quoteArgs": false
            },
            "processId": "${command:pickRemoteProcess}",
            "sourceFileMap": {
                "/app": "${workspaceRoot}"
            },
            "justMyCode": true
        }
    ]
}
```

`resolveDebugConfigurationWithSubstitutedVariables` does not work for this use case as both commands are resolved/executed before replacement occurs.  `${command:pickRemoteProcess}` needs the value from `${command:vscode-docker.containers.select}` in the config object passed to it during execution.

### CODE REVIEW

**Comment by weinand** [comment]
@davidruhmann what is exactly the approach behind your change?
What would we add to the variables documentation to explain the behavior to users?

**Comment by davidruhmann** [comment]
Definitely.  Appreciate your taking a look at this.

My approach to the change was to be as backwards compatible as possible and not change the input or output of the public functions.

All the changes culminate around making this line pass in an object with the known replacements inserted.
```
result = await this.commandService.executeCommand(commandId, object);
```

So when this `${command:pickRemoteProcess}` command is executed, the previous command's value would already be replaced in the configuration passed in since the replacement is done sequentially.
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Docker Attach",
            "type": "coreclr",
            "request": "attach",
            "pipeTransport": {
                "pipeProgram": "docker",
                "pipeArgs": [
                    "exec", "-i", "docker-select-value" // Actual value for pickRemoteProcess to use
                ],
                "debuggerPath": "/root/vsdbg/vsdbg",
                "pipeCwd": "${workspaceRoot}",
                "quoteArgs": false
            },
            "processId": "${command:pickRemoteProcess}",
            "sourceFileMap": {
                "/app": "${workspaceRoot}"
            },
            "justMyCode": true
        }
    ]
}
```

Previously, the object passed into the command execution would be the raw original without replacement values.
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Docker Attach",
            "type": "coreclr",
            "request": "attach",
            "pipeTransport": {
                "pipeProgram": "docker",
                "pipeArgs": [
                    "exec", "-i", "${command:vscode-docker.containers.select}" // Value is known in the array, but not replaced till after all other commands are executed.
                ],
                "debuggerPath": "/root/vsdbg/vsdbg",
                "pipeCwd": "${workspaceRoot}",
                "quoteArgs": false
            },
            "processId": "${command:pickRemoteProcess}",
            "sourceFileMap": {
                "/app": "${workspaceRoot}"
            },
            "justMyCode": true
        }
    ]
}
```


As for documentation updates, I do not think any updates are needed; as the current documentation makes it sound like how it works with my changes is how it should work.
https://code.visualstudio.com/docs/editor/variables-reference#_command-variables
> When a command variable is interpolated, the command is run and the variable is substituted by the command's (string) result.


Please let me know if I have missed something or need to update anything else.

File: src/vs/workbench/services/configurationResolver/browser/configurationResolverService.ts [hunk 1]
```diff
@@ -164,12 +165,9 @@ export abstract class BaseConfigurationResolverService extends AbstractVariableR
 		}
 
 		// extract and dedupe all "input" and "command" variables and preserve their order in an array
-		const variables: string[] = [];
-		this.findVariables(configuration, variables);
-
 		const variableValues: IStringDictionary<string> = Object.create(null);
-
-		for (const variable of variables) {
+		const config = deepClone(configuration);
```
Comment (davidruhmann) [inline src/vs/workbench/services/configurationResolver/browser/configurationResolverService.ts]: Make a clone of the configuration so we do not manipulate the object passed into this function.

File: src/vs/workbench/services/configurationResolver/browser/configurationResolverService.ts [hunk 2]
```diff
@@ -184,7 +182,7 @@ export abstract class BaseConfigurationResolverService extends AbstractVariableR
 				case 'command':
 					// use the name as a command ID #12735
 					const commandId = (variableToCommandMap ? variableToCommandMap[name] : undefined) || name;
-					result = await this.commandService.executeCommand(commandId, configuration);
+					result = await this.commandService.executeCommand(commandId, object);
```
Comment (davidruhmann) [inline src/vs/workbench/services/configurationResolver/browser/configurationResolverService.ts]: The object passed into executeCommand now has all the previously evaluated replacements inserted.

File: src/vs/workbench/services/configurationResolver/browser/configurationResolverService.ts [hunk 3]
```diff
@@ -197,46 +195,69 @@ export abstract class BaseConfigurationResolverService extends AbstractVariableR
...
+			const input = object;
+			while ((matches = BaseConfigurationResolverService.INPUT_OR_COMMAND_VARIABLES_PATTERN.exec(input)) !== null) {
 				if (matches.length === 4) {
+					const entry = matches[0];
 					const command = matches[1];
-					if (variables.indexOf(command) < 0) {
-						variables.push(command);
+					const found = command in variables;
+					if (!found) {
+						// NOTE this if assumes that the user will only ever want one computed value. (e.g. have two of the same command but only get the first output for both)
```
Comment (davidruhmann) [inline src/vs/workbench/services/configurationResolver/browser/configurationResolverService.ts]: This assumption was from the existing code.  Just added a comment to make it clear while I was making changes.

**Comment by davidruhmann** [comment]
@weinand 
Interesting point that I had forgotten about on the json ordering non-guarantee.  Looking at some different ways to resolve that.  Now that you mention it, I think I remember seeing some discussion on ordering of variables/commands from a while back in an issue.  Going to see if I can find the discussion...

Also have some ideas on how to resolve the ordering that I will play with.  Do you have any specific thoughts around it?

Some of my initial thoughts:
* Order by nested level.  Guarantee variables in a nested level object are resolved before the parent variables.  This would not guarantee ordering of variables at the same level.  This would just be an initial solution that requires no user changes.
* Add ordering notation to the variables (bigger change).  (Think this is what the discussion I am trying to find was about)

**Comment by weinand** [comment]
@davidruhmann I've reopened issue #89758 

Let's continue the discussion there and please verify that my [take of the problem](https://github.com/microsoft/vscode/issues/89758#issuecomment-599561497) is correct.

**Comment by weinand** [comment]
@davidruhmann Thanks a lot for the PR, but we do not plan to solve #89758 on the VS Code side. Please see my [comment]( https://github.com/microsoft/vscode/issues/89758#issuecomment-721630063)

**Pull request closed without merging**